Agent investment (part 1)

alexdewar opened the pull request:
# Description

I thought about breaking this up into smaller PRs, but tbh I think it might be easier to review in one go. I can break it up a bit if you'd find it easier, though.

This PR adds code for agent investment up to the point where assets are selected and added to the asset pool for the next year. It doesn't go as far as running dispatch for the next milestone year or the "ironing out" loop. Only SVD commodities are considered for now (see #650).

I changed the API for `AssetPool` again :upside_down_face:. The reason is because we will want to include assets selected in prior iterations of the agent investment loop (for the same milestone year) in the dispatch optimisation at some point and we *could* end up with different, non-commissioned assets having the same hash, which will break things. So I added an `extend` method to `AssetPool` so we can grow it incrementally and commission assets as we go along (commissioned assets will all have different hashes because they have different IDs).

The current workflow for agent investment is:

1. For every commodity, iterate over the agents responsible for a portion of its production
2. Iterate over the regions in which the current agent operates
3. Calculate the maximum capacity for candidate assets (see #693)
4. Get existing and candidate assets
5. Select the best assets from among these (using NPV/LCOX) to meet required demand and add to asset pool

Closes #693. Closes #690. Closes #687.

## Type of change

- [ ] Bug fix (non-breaking change to fix an issue)
- [x] New feature (non-breaking change to add functionality)
- [ ] Refactoring (non-breaking, non-functional change to improve maintainability)
- [ ] Optimization (non-breaking change to speed up the code)
- [ ] Breaking change (whatever its nature)
- [ ] Documentation (improve or add documentation)

## Key checklist

- [x] All tests pass: `$ cargo test`
- [ ] The documentation builds and looks OK: `$ cargo doc`

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

alexdewar commented on src/simulation/investment.rs:
```diff
@@ -67,6 +128,242 @@ pub fn get_demand_profile(
+            (
+                time_slice.clone(),
+                *demand
+                    .get(&(commodity_id.clone(), region_id.clone(), time_slice.clone()))
```
This is wrong

alexdewar commented on src/simulation/investment.rs:
```diff
@@ -67,6 +128,242 @@ pub fn get_demand_profile(
+                .get(&(commodity_id.clone(), region_id.clone(), time_slice.clone()))
+                .unwrap()
+        })
+        .max_by(|a, b| a.total_cmp(b))
```
`Flow` does not implement `Ord` :stuck_out_tongue:. But it could do, if we think it's worth it.

There is the [`ordered_float`](https://docs.rs/ordered-float/latest/ordered_float/) crate we could use if we don't want to implement it ourselves.

alexdewar commented:
Rebasing to get fixed links in #708

tsmbland commented on src/simulation/investment.rs:
```diff
@@ -67,6 +128,242 @@ pub fn get_demand_profile(
+    region_id: &RegionID,
+) -> Capacity {
+    model.parameters.capacity_limit_factor
+        * get_peak_demand(&model.time_slice_info, demand, commodity_id, region_id)
```
I may well have misunderstood, but this seems a bit different from how I interpreted this parameter (based on https://github.com/EnergySystemsModellingLab/MUSE_2.0/issues/687#issuecomment-3049316024)

For a start, I would say you need to calculate this separately for each candidate asset, as it seems strange to use the same capacity limit for different candidate assets regardless of whether the output coefficient for the commodity of interest is 100 or 0.001. I'd say you'd require a higher capacity limit for the latter (or maybe not - correct me if I'm wrong!!). This becomes a bit complicated when you take into account activity limits, which may also differ by timeslice.

Let's think of an example (in this case a model with just two timeslices):

Demand for commodity X in each timeslice (`Flow`) = [5,6]
Upper activity per capacity limits (`ActivityPerCapacity`) = [1,2]
Flow coefficients for commodity X (`FlowPerActivity`) = [3,4]
-> Upper flow limits for commodity X (`FlowPerCapacity`) = [3, 8]
-> Capacity required to meet demand in each timeslice (`Capacity`) = [5/3, 6/8]

In MUSE1 we would use 5/3 as the upper capacity limit, as this is the capacity that would be able to meet demand in all timeslices and no more (this is the "demand limiting capacity" constraint). But I'm not quite sure what we should be doing here with this new "capacity_limit_factor".

@alexdewar @ahawkes

ahawkes commented:
If I understand correctly, this relates to determining what capacity limit to use when doing an investment appraisal (and in the MVP we are only implementing the annualised profit maximising approach)?

There probably is not “right” answer here.

Suggest we go with the interpretation that it should be ability to output the commodity of interest. So that would be capacity*cap2act*coeff_coi?


From: Tom Bland ***@***.***>
Date: Monday, 21 July 2025 at 17:29
To: EnergySystemsModellingLab/MUSE_2.0 ***@***.***>
Cc: Hawkes, Adam D ***@***.***>, Mention ***@***.***>
Subject: Re: [EnergySystemsModellingLab/MUSE_2.0] Agent investment (part 1) (PR #707)

@tsmbland commented on this pull request.

________________________________

In src/simulation/investment.rs<https://github.com/EnergySystemsModellingLab/MUSE_2.0/pull/707#discussion_r2219685374>:

> +    agents.filter(move |agent| {

+        agent

+            .commodity_portions

+            .contains_key(&(commodity_id.clone(), year))

+    })

+}

+

+/// Get the maximum candidate asset capacity

+fn get_maximum_candidate_capacity(

+    model: &Model,

+    demand: &AllDemandMap,

+    commodity_id: &CommodityID,

+    region_id: &RegionID,

+) -> Capacity {

+    model.parameters.capacity_limit_factor

+        * get_peak_demand(&model.time_slice_info, demand, commodity_id, region_id)

I may well have misunderstood, but this seems a bit different from how I interpreted this parameter (based on #687 (comment)<https://github.com/EnergySystemsModellingLab/MUSE_2.0/issues/687#issuecomment-3049316024>)

For a start, I would say you need to calculate this separately for each candidate asset, as it seems strange to use the same capacity limit for different candidate assets regardless of whether the output coefficient for the commodity of interest is 100 or 0.001. I'd say you'd require a higher capacity limit for the latter (or maybe not - correct me if I'm wrong!!). This becomes a bit complicated when you take into account activity limits, which may also differ by timeslice.

Let's think of an example (in this case a model with just two timeslices):

Demand for commodity X in each timeslice (Flow) = [5,6]
Upper activity per capacity limits (ActivityPerCapacity) = [1,2]
Flow coefficients for commodity X (FlowPerActivity) = [3,4]
-> Upper flow limits for commodity X (FlowPerCapacity) = [3, 8]
-> Capacity required to meet demand in each timeslice (Capacity) = [5/3, 6/8]

In MUSE1 we would use 5/3 as the upper capacity limit, as this is the capacity that would be able to meet demand in all timeslices and no more (this is the "demand limiting capacity" constraint). But I'm not quite sure what we should be doing here with this new "capacity_limit_factor".

@alexdewar<https://github.com/alexdewar> @ahawkes<https://github.com/ahawkes>

—
Reply to this email directly, view it on GitHub<https://github.com/EnergySystemsModellingLab/MUSE_2.0/pull/707#pullrequestreview-3038851034>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AC37JLKRAD7L76GHOXRXHMT3JUIMTAVCNFSM6AAAAACBYB55MOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMZYHA2TCMBTGQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

alexdewar commented:
Good point @tsmbland.

@ahawkes:

> If I understand correctly, this relates to determining what capacity limit to use when doing an investment appraisal (and in the MVP we are only implementing the annualised profit maximising approach)?

Yep. (Although in the end, @tsmbland implemented NPV and LCOX, because they're fairly similar.)

> There probably is not “right” answer here.
>
> Suggest we go with the interpretation that it should be ability to output the commodity of interest. So that would be capacity\*cap2act\*coeff_coi?

I'm guessing the `capacity` here is the max capacity I've already calculated as `(peak demand for COI) * (capacity limit factor)` (where "capacity limit factor" is the new parameter we talked about, defaulting to 0.1)?

NB: We will end up with a unit type mismatch here, in that the result will be of type `Flow` not `Capacity`, but we can cast it back to `Capacity`, as long as the maths makes sense.

Something else that's occurred to me is that we're not taking into account time slice length here, which seems a bit suspect. You may end up finding "peak demand" in a given time slice just because it happens to be very long. Would it be better to use peak load?

tsmbland commented:
> Suggest we go with the interpretation that it should be ability to output the commodity of interest. So that would be capacity*cap2act*coeff_coi?

@ahawkes To clarify, are you suggesting we don't take into account the upper activity limit of the process (i.e. what we call utilisation factor in MUSE1)?

> Something else that's occurred to me is that we're not taking into account time slice length here, which seems a bit suspect. You may end up finding "peak demand" in a given time slice just because it happens to be very long. Would it be better to use peak load?

@alexdewar True, as it's currently done. But activity limits are calculated taking timeslice lengths into account ([here](https://github.com/EnergySystemsModellingLab/MUSE_2.0/blob/504cda09d4cbd515be5e442ffc0310baef73028c/src/input/process/availability.rs#L43)), so in the example I gave above this is all fine. No problems with units either

tsmbland commented:
> No no – the upper activity limit is also taken account of. Apologies missed that bit. capacity*activity_upper_limit_factor*cap2act*coeff_coi (and dealing with time slice length)

Ok great. Then the issue is that activity limits can vary by timeslice, so I'd suggest we do this for all timeslices and take the maximum (as in the above example). Then multiply this by the `capacity_limit_factor`. Pretty sure the units will all work out fine. @alexdewar makes sense?

alexdewar commented:
> Sure, but if the units of `capacity_limit_factor` aren't `Dimensionless` then I think we've done something wrong

Oh yeah :laughing:. Let's see how it looks

alexdewar commented:
Think this is ready for review now @tsmbland. I've changed the calculation for max capacity as discussed and added a couple of tests.

alexdewar commented on src/simulation/investment.rs:
```diff
@@ -67,15 +129,311 @@ pub fn get_demand_profile(
+        // Adjust for cap2act and time slice duration
+        let max_act_per_cap = activity_upper
+            * process.parameters[&(region_id.clone(), year)].capacity_to_activity
+            / Dimensionless(duration.value());
```
Oops :see_no_evil:. That's embarrassing...

Am I right in thinking we *could* conceptualise the activity limits as `ActivityPerYear`? This would mean that `process.activity_limits` used units of `PerYear`, which is admittedly a bit odd, but I *think* would be consistent.

tsmbland replied:
> Am I right in thinking we _could_ conceptualise the activity limits as `ActivityPerYear`? This would mean that `process.activity_limits` used units of `PerYear`, which is admittedly a bit odd, but I _think_ would be consistent.

It's possible, so essentially we'd be describing a limit on the _rate of activity_ over a time period, rather than absolute activity within a time period (the former being insensitive to the duration of the time period). 

I can't really say whether this would help or not without giving it a try, but my suspicion is that it would be very fiddly to change, might overcomplicate things, and probably wouldn't help much anyway (e.g. you'd still need to work out from context whether the rate applies to an individual timeslice or the year as a whole)

alexdewar commented on src/simulation/investment.rs:
```diff
@@ -136,4 +494,175 @@ mod tests {
+        commodity_id: CommodityID,
+    ) {
+        // Setup two time slices
+        let ts_day = TimeSliceID {
```
There is, but Copilot didn't spot it :stuck_out_tongue:. I'll fix it up

alexdewar commented:
> Thanks for explaining. It's probably fine as it is then, but if `ordered_float` is easy to implement then it might be nice.

I'll take another look tomorrow to see how faffy it is. Maybe it should be a separate PR.

> Worth thinking about nans though. I don't imagine we'll ever want unit types to be nan, but there are certainly times where these could crop up and we'll need to deal with them. I wonder if [`NotNan`](https://docs.rs/ordered-float/latest/ordered_float/struct.NotNan.html) could be something to consider. This will panic if nans are ever encountered, which will force us to explicitly deal with these scenarios (e.g. adding an epsilon to prevent a 0/0 division)

Yeah, I don't think so either. I think it would be a bit more Rusty to have an `Option<Activity>` rather than an `Activity` that might be NaN. I've grepped the code base and it looks like the only NaN unit types that exist are in tests, so it *might* be safe to swap over.

My only worry is that there could be performance implications to the extra checks that would be implicitly needed for all kinds of floating point operations, but maybe that's not worth worrying about.

The other thing is that we would need to create unit types with a constructor, because writing e.g. `Activity(1.0)` wouldn't work anymore.

alexdewar commented:
I'll leave the units thing for now as it will require a bit of refactoring. I think I've addressed the other feedback.

tsmbland commented:
> According to the docs the coefficient for the capacity var should be A F C for LCOX and − A F C for NPV, but it's currently set to `1.0` in both cases... I guess you were planning on doing this later @tsmbland?

I think this is all correct in the code. What are you looking at?

ahawkes commented:
It is possible that no asset is profitable - ProfitPerAct is always negative - so the profit maximisation solution is to do nothing. I think we would need to see values of the parameters going into this for a simple example, and the ProfitPerAct value, to understand the behaviour. Also possible that assets are profitable, but for some signing convention reason ProfitPerAct should be multiplied by -1.0.

Is it possible to see some values?

alexdewar commented on src/simulation/investment.rs:
```diff
@@ -67,15 +128,282 @@ pub fn get_demand_profile(
+            .unwrap();
+
+        // Remove this capacity from the available remaining capacity for this asset
+        old.make_mut().capacity -= capacity;
```
Ahhh yes, that makes sense. We need to separate out the two. I'll have a go. Maybe it will fix the infinite loop!

alexdewar replied:
I've had a go at implementing this

alexdewar commented on src/simulation/investment.rs:
```diff
@@ -67,15 +123,287 @@ pub fn get_demand_profile(
+
+        let mut current_best: Option<AppraisalOutput> = None;
+        for asset in opt_assets.iter() {
+            let max_capacity = (!asset.is_commissioned()).then(|| {
+                let max_capacity = model.parameters.capacity_limit_factor * asset.capacity;
+                let remaining_capacity = remaining_candidate_capacity[asset];
+                max_capacity.min(remaining_capacity)
+            });
```
I guess an alternative would be to define a separate `Asset::is_candidate` method for clarity

alexdewar commented:
Right, I think this is good to go now :partying_face:. I've merged in `main` and fixed it up.

I made some small changes to the appraisal code:

- I added a separate (optional) `max_capacity` arg so that candidates can specify the upper limit without having to create a new `AssetRef` with the desired capacity
- Refactor to reduce the number of arguments so we don't fall foul of clippy's `too_many_arguments` lint
